modules/inject/src/main/scala/io/janstenpickle/trace4cats/inject/ContextConstructor.scala

```diff
@@ -0,0 +1,51 @@
+package io.janstenpickle.trace4cats.inject
+
+import cats.Applicative
+import cats.effect.Resource
+
+trait ContextConstructor[F[_], R, R0] { self =>
+  def F: Applicative[F]
+
+  def apply(r: R): Resource[F, R0] = resource(r)
+
+  def resource(r: R): Resource[F, R0]
+
+  def mapBoth[R1](f: (R, R0) => R1): ContextConstructor[F, R, R1] = new ContextConstructor[F, R, R1] {
```

Comment:
Here and below I see naming issues.
Let's call its methods according to the common FP notation.

We could derive some instances: `Functor[ContextConstructor[F[_], In, *]]`, `Contravariant[ContextConstructor[F[_], *, Out]]`, `Arrow[ContextConstructor[F[_], *, *]`, so let's use the same names as defined in those typeclasses.